spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/KerasLayer.scala

```diff
@@ -0,0 +1,243 @@
+    }
+    if (shape.isInstanceOf[SingleShape]) {
+      Shape((List(-1) ++ shape.toSingle()).toArray)
+    } else {
+      Shape(shape.toMulti().map {addBatch(_)})
+    }
+  }
+
+  def removeBatch(shape: Shape): Shape = {
+    if (shape == null) {
+      return null
+    }
+    if (shape.isInstanceOf[SingleShape]) {
+      Shape((shape.toSingle().slice(1, shape.toSingle().length)).toArray)
+    } else {
+      Shape(shape.toMulti().map {addBatch(_)})
+    }
+  }
+}
+
+/**
+ * KerasModule is the basic component of all Keras-like Layer.
+ * It forward activities and backward gradients, and can be mixed with other AbstractMoudule.
+ *
+ * @tparam A Input data type
+ * @tparam B Output data type
+ * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now
+ * @param batchInputShape the first dim is batch
+ */
+abstract class KerasLayer[A <: Activity: ClassTag, B <: Activity: ClassTag, T: ClassTag]
+(batchInputShape: Shape = null)(implicit ev: TensorNumeric[T]) extends Container[A, B, T]{
+
+  def labor: AbstractModule[A, B, T] = {
+    if (this.modules.isEmpty) {
+      throw new RuntimeException("This Layer hasn't been built")
+    }
+    require(modules.length == 1,
+      s"modules should only contain 1 element instead of ${modules.length}")
+    modules(0).asInstanceOf[AbstractModule[A, B, T]]
+  }
+
+  // scalastyle:off
+  def labor_=(value: AbstractModule[A, B, T]): Unit = {
+    modules.clear()
+    modules.append(value)
+  }
+  // scalastyle:on
+
+  override def output: B = labor.output
+
+  override def gradInput: A = labor.gradInput
+
+  override def inputShapeValue: Shape = labor.inputShapeValue
+
+  override def outputShapeValue: ArrayBuffer[Shape] = labor.outputShapeValue
+
+  // scalastyle:off
+  override def output_= (value: B): Unit = labor.output = value
+
+  override def gradInput_=(value: A): Unit = {
+    labor.gradInput = value
+  }
+
+  override def inputShapeValue_=(value: Shape): Unit = {
+    labor.inputShapeValue = value
+  }
+
+  override def outputShapeValue_=(value: ArrayBuffer[Shape]): Unit = {
+   labor.outputShapeValue = value
+  }
+  // scalastyle:on
+
+  override def updateOutput(input: A): B = {
+    labor.updateOutput(input)
+  }
+
+  def updateGradInput(input: A, gradOutput: B): A = {
+    labor.updateGradInput(input, gradOutput)
+  }
+
+  override def isCompatibleWithKeras(): Boolean = true
+
+  override def isCompatibleWithTorch(): Boolean = false
+
+  override def getInputShape(): Shape = {
+    if (batchInputShape != null) {
+      batchInputShape
+    } else if (this.labor == null) {
+      null
+    } else {
+      this.labor.getInputShape()
+    }
+  }
+
+  override def computeOutputShape(inputShape: Shape): Shape = {
+    this.labor.computeOutputShape(inputShape)
+  }
+
+  override def getOutputShape(): Shape = labor.getOutputShape()
+
+  override def build(inputShape: Shape): Shape = {
+    labor = doBuild(inputShape)
+    if (!labor.isBuilt) {
+      labor.build(inputShape)
```

Comment:
Did you fix the problem of putting `computeOutputShape` logic in new layers?
Currently in your implementation, after calling `doBuild`, we will get `labor` and `labor.build` will be invoked, which means that we still need `computeOutputShape` in labor? Then when will the overridden method in the new layer be called? In `super.build(inputShape)`? But before that, if the old layer doesn't implement outputshape, an exception is already there.......